### PR TITLE
Config and serialization enhancement

### PR DESCRIPTION
**Pull request by AdrianSosic**
This PR significantly enhances our serialization machinery:
* Search spaces can now be deserialized using a selected classmethod. Currently, this is enabled by manually registering hooks for our search space classes. However, if we add more constructors to other classes in the future, we can consider automating the process using hook predicate functions / factories. Perhaps, there will even be a dedicated [strategy](https://github.com/python-attrs/cattrs/issues/489) available for this mechanism.
* Due to this change, there is no more separate "config converter" required, since creation from config can now happen via the regular converter. In fact, in future we can even think about deprecating our "from_config" approach, because there really is no more such thing as a "config" – it's just a "regular" JSON string that goes through the default converter.
* The serialization functionality now sits in its own subpackage.
* Missing serialization mixins have been added to `Interval` and the two subspace classes.
* Added a basic serialization roundtrip test for dataframes and a corresponding hypothesis strategy.
* Changed binarization of dataframes to use regular pickle due to some edge cases with the previous `parquet` approach that where detected through the above test.
* Added a `SearchSpace.from_dataframe` convenience constructor for consistency, which is also helps to simplify campaign configs.

### CODE REVIEW

**Comment by AdrianSosic** [comment]
Hi @AVHopp, could you do me a favor? The doc build is currently failing due to an apparently unresolved reference. However, the error message is not really helpful and I have no clue what is the problem. Perhaps you have an immediate idea when you look at the logs?

File: baybe/serialization/mixin.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import json
+from typing import Type, TypeVar
+
+from baybe.serialization.core import converter
+
+_T = TypeVar("_T")
+
+
+class SerialMixin:
```
Comment (AVHopp) [inline baybe/serialization/mixin.py:11]: [Putting this here since there is no better place]
**TL;DR:** Change line 
image
in `conf.py` to 
image

to fix doc building issue. Tested locally, works.

**Explanation:** `sphinx` tries to retrieve the init function of the classes for the autodocumentation. However, we have some rare instances where we do not use `attrs` for automatically creating them and do not provide them ourselves. For this reason, we need to explicitly tell sphinx "Please ignore the following functions" which is done by the `nitpck_ignore_regex` list in `conf.py`. Since the structure of the class changed, we need to change the part here as well.

This issue with outdated links happen in some more places (which is only visible after this fix). I'll create two commits to fix the corresponding issues.

Comment (AVHopp) [reply baybe/serialization/mixin.py:11]: Corresponding commits are pushed, so feel free to resolve if docs now works and this explanation makes sense.

Comment (AdrianSosic) [reply baybe/serialization/mixin.py:11]: Nice, thanks for the fix 🏅 
Will rebase the branch on main to resolve the current merge conflicts and then we should see the pipeline pass!

File: baybe/serialization/core.py [hunk 2]
```diff
@@ -111,19 +65,15 @@ def structure_base(val: dict, _: Type[_T]) -> _T:
 
 
 def _structure_dataframe_hook(string: str, _) -> pd.DataFrame:
-    """De-serialize a DataFrame."""
-    buffer = BytesIO()
-    buffer.write(base64.b64decode(string.encode("utf-8")))
-    return pd.read_parquet(buffer)
+    """Deserialize a DataFrame."""
+    pickled_df = base64.b64decode(string.encode("utf-8"))
+    return pickle.loads(pickled_df)
```
Comment (AVHopp) [inline baybe/serialization/core.py:70]: What is the reason for using pickle now?

**Review by AVHopp — APPROVED** [review]
Didn't spot anything that needs a change. Note however that I am not hat familiar with all of the serialization stuff, so there is a chance I missed stuff.
If I should give a specific part a more in-depth treatment just tell me.

File: baybe/campaign.py [hunk 3]
```diff
@@ -119,19 +112,21 @@ def from_config(cls, config_json: str) -> Campaign:
         Returns:
             The constructed campaign.
         """
+        from baybe.deprecation import compatibilize_config
```
Comment (Scienfitz) [inline baybe/campaign.py:115]: why not top level import?

File: baybe/campaign.py [hunk 4]
```diff
@@ -119,19 +112,21 @@ def from_config(cls, config_json: str) -> Campaign:
...
-        }
-        return _config_converter.structure(config, Campaign)
+
+        # Temporarily enable backward compatibility
+        config = compatibilize_config(config)
+
+        return converter.structure(config, Campaign)
 
     @classmethod
     def to_config(cls) -> str:
```
Comment (Scienfitz) [inline baybe/campaign.py:125]: how do you envision `to_config` in the context of the new changes? There ar eprob many problems to solve. To is `to_config` still reasonable if `from_config` somehow gets absorbed by `from_json`? Why do we even have this unimplemented function already included here?

File: baybe/searchspace/continuous.py [hunk 5]
```diff
@@ -69,20 +70,20 @@ def from_bounds(cls, bounds: pd.DataFrame) -> SubspaceContinuous:
         return SubspaceContinuous(parameters)
 
     @classmethod
-    def from_dataframe(cls, df: pd.DataFrame) -> SubspaceContinuous:
+    def from_dataframe(cls, dataframe: pd.DataFrame) -> SubspaceContinuous:
```
Comment (Scienfitz) [inline baybe/searchspace/continuous.py:73]: `dataframe` is just a long version of the very common `df`, why rename? thats prob the most commonly used convention in all of data science python :O

File: baybe/searchspace/discrete.py [hunk 6]
```diff
@@ -206,6 +208,12 @@ def from_dataframe(
         Raises:
             ValueError: If several parameters with identical names are provided.
             ValueError: If a parameter was specified for which no match was found.
+
+        Note:
+            For columns that have no corresponding entry provided via `parameters`,
```
Comment (Scienfitz) [inline baybe/searchspace/discrete.py:213]: is this not redundant with the docstring already present for `parameters` above?

in a way this is telling me `if not specified, defaults will be used`... isnt that obvious? 
one could also live without the note by simply adding `(with default arguments)` to the docstring above

File: baybe/searchspace/discrete.py [hunk 7]
```diff
@@ -382,3 +390,7 @@ def parameter_cartesian_prod_to_df(
     ret = pd.DataFrame(index=index).reset_index()
 
     return ret
+
+
+# Register deserialization hook
+converter.register_structure_hook(SubspaceDiscrete, select_constructor_hook)
```
Comment (Scienfitz) [inline baybe/searchspace/discrete.py:396]: so this constructor hook seems rather general, is there any reason why its only enabled for the searchspace and not immediately for every object?

File: baybe/searchspace/core.py [hunk 8]
```diff
@@ -146,6 +146,50 @@ def from_product(
...
+        if set(p.name for p in parameters) != set(dataframe.columns.values):
+            raise ValueError(
+                "The provided dataframe columns must match exactly with the specified "
+                "parameter names."
+            )
+
+        disc_params = [p for p in parameters if p.is_discrete]
+        cont_params = [p for p in parameters if not p.is_discrete]
+
+        # TODO: The continuous classmethod does not yet accept a parameter argument.
```
Comment (Scienfitz) [inline baybe/searchspace/core.py:183]: thats a small change, I suggest to either include it or remove this todo (why do conti params need this argument?)

File: baybe/serialization/core.py (hunk 2 above)
Comment (AdrianSosic) [reply baybe/serialization/core.py:70]: Simply because that whole dataframe serialization topic is just a nightmare 🙈 We started using human-readable stuff like csv but always got type inference problems in roundtrip serialization (e.g. ensure that a float remains a float after de-/serializing and is not converted to int, even when it's numeric value is 0.0). That's why – for the lack of better alternatives – I switched to binary representation at some point. But now the hypothesis test also discovered some edge cases for the parquet format I used ... All in all, this problem is not yet fully solved and will probably undergo more changes in the future, until we find a version that is both robust and ideally human-readable. So for now, just consider it as a quick fix, but feel free to look for better alternatives 🙃 Will resolve, but reactivate if further discussion is needed.

File: examples/Serialization/create_from_config.py [hunk 9]
```diff
@@ -20,50 +20,53 @@
...
-    "parameters": [
-        {
-            "type": "CategoricalParameter",
-            "name": "Granularity",
-            "values": [
-                "coarse",
-                "fine",
-                "ultra-fine"
+    "searchspace": {
+        "constructor": "from_product",
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply examples/Serialization/create_from_config.py:24]: Fixed, thanks.

File: examples/Serialization/create_from_config.py [hunk 10]
```diff
@@ -20,50 +20,53 @@
...
-        {
-            "type": "NumericalDiscreteParameter",
-            "name": "Pressure[bar]",
-            "values": [
-                1,
-                5,
-                10
-            ],
-            "tolerance": 0.2
+        "objective": {
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply examples/Serialization/create_from_config.py:61]: fixed

File: baybe/serialization/utils.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
+"""A collection of serialization utilities."""
+
+from typing import Any
+
+import pandas as pd
+
+
+def serialize_dataframe(df: pd.DataFrame, /) -> Any:
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/serialization/utils.py:8]: No, it's really just to encapsulate the logic gives you the required serialization string for the JSON config. Although we both agree that the current binary approach is not yet optimal, we should (at least temporarily) expose some public method that takes care of the conversion, I think. Otherwise, I'd need to tell my colleagues who use the config "to get your dataframe string, import that one custom converter, bla bla ..." <--- not so nice. Are you OK with this or do you have a better idea?

File: baybe/searchspace/continuous.py (hunk 5 above)
Comment (AdrianSosic) [reply baybe/searchspace/continuous.py:73]: I know 😃 And we can also stick with `df` if you prefer, depending on what you think about that one issue that you might not have considered: in the current setting, the name of the parameter is what you need to provide in your config, i.e. you'd need to write
```json
"constructor": "from_dataframe",
"df": ...
```
That was my reason for the change. If you think that is better than the current version, I'm happy to undo

File: baybe/searchspace/core.py (hunk 8 above)
Comment (AdrianSosic) [reply baybe/searchspace/core.py:183]: Was mostly there to raise your attention, which seems to have worked 😄 I just noticed that slight inconsistency in the signatures of both methods. For discrete parameters, we certainly need the explicit parameter specification (e.g. to be able to fully configure them, explicitly choose their types, etc.). For conti parameters, this is not strictly needed at the moment because there is only one type available. However, while writing, I just noticed that we'd also need the argument if we want to explicitly set bounds for some of the parameters. So I guess I'll just add it, what do you think?

File: baybe/campaign.py (hunk 3 above)
Comment (AdrianSosic) [reply baybe/campaign.py:115]: Classic: circular import. For this particular one I wouldn't even care too much since it's deprecation code anyway. However, perhaps we can at some point do one iteration and see if we can remove some of the delayed imports at other places.... For now, I'll resolve though

File: baybe/searchspace/discrete.py (hunk 7 above)
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:396]: Upfront: have you seen my note in the PR description? I think this not yet the final version but will also depend on how it's going to be included into cattrs. But to answer the question: simply because it would be inefficient. Following that same logic, we could also have a general "if object == my_type" check at the beginning of all hooks we write and then apply it to ALL python objects instead of just to "my_type" – you know what I mean? The hook would be called for no reason. I think the clean way would be to either manually register it to exactly the types we need (current approach) or perhaps use some predicate function to decide, but I wouldn't bother until the question what happens on cattrs side is fully answered. Will draft a cattrs PR for it and then we'll see. Please reactivate if you want to further discuss.

File: baybe/campaign.py (hunk 4 above)
Comment (AdrianSosic) [reply baybe/campaign.py:125]: Yeah,  agree, `to_config` has been hanging around unimplemented way too long already. Let me remove it in this PR and we can bring it back once actually implemented. Concerning its responsibility: we can definitely discuss it! The way I see it is that it basically gives you a minimal config back, instead of a full serialized object. In a sense, it would close the loop for a roundtrip that starts from the serialized side. For example, I you currently create a campaign using `from_product`, you cannot recreate that config from the object currently, because `to_json` would yield the expanded search space object. Having a dedicated `to_config` method would allow to create these minimal configs from existing python objects. Will create a workitem and resolve.

File: baybe/searchspace/discrete.py (hunk 6 above)
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:213]: Hm, not sure actually. The reason why I added it is because I just thought about it for the first time when I tested roundtrip serialization and noticed that the optional arguments were (of course) not set. And this is even though I was the one who wrote the method 😄 @AVHopp: what do you think? Is the note helpful or unnecessary since clear from the context?

File: baybe/serialization/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/serialization/utils.py:8]: ok argument makes sense --> provide a utility for people

File: baybe/searchspace/continuous.py (hunk 5 above)
Comment (Scienfitz) [reply baybe/searchspace/continuous.py:73]: ok didnt thing about that, true
but in my midn `df` is still as clear as `dataframe` and would break the code consistency? People will have to look up constructor signatures anyway, so why bother with this?

but ill leave it to you, maybe @AVHopp has an opinion, feel free to resolve

File: baybe/searchspace/core.py (hunk 8 above)
Comment (Scienfitz) [reply baybe/searchspace/core.py:183]: yep agree

File: baybe/searchspace/discrete.py [hunk 12]
```diff
@@ -182,14 +183,15 @@ def from_product(
     @classmethod
     def from_dataframe(
         cls,
-        df: pd.DataFrame,
+        dataframe: pd.DataFrame,
         parameters: Optional[List[Parameter]] = None,
```
Comment (Scienfitz) [inline baybe/searchspace/discrete.py:187]: shouldnt this be `DiscreteParameter` ?

File: baybe/searchspace/continuous.py (hunk 5 above)
Comment (AVHopp) [reply baybe/searchspace/continuous.py:73]: My feeling is also that this is overthinking. And also, we have type hints for telling people what they actually need to provide, so there is no ambiguity here (assuming that people read the type hints which I think we should assume). So my vote is to go for `df` instead of dataframe but keep the constructor name.

File: baybe/searchspace/discrete.py (hunk 6 above)
Comment (AVHopp) [reply baybe/searchspace/discrete.py:213]: I understand the note basically as "If you do not provide additional arguments, defaults are used", so this should be clear in general. I however also see the point that it might make sense to make people aware of that.
I'd suppose to get rid of the second sentence (that feels really redundant to me) and include the note about the default values in the `Args` section for `parameters`. If that makes the `Args` section too packed, I'd also rather remove the note.

**Review by Scienfitz — APPROVED** [review]

File: baybe/searchspace/core.py (hunk 8 above)
Comment (AdrianSosic) [reply baybe/searchspace/core.py]: Done. Have also extracted a utility to avoid code duplication for both subspaces. Along the way, have fixed the types and also rewritten the docstrings. Most importantly, have added tests for all three constructors to ensure they do what they are supposed to do. Hope that all open issues are resolved now. Can you have a look at the last few commits and confirm?

File: baybe/searchspace/discrete.py (hunk 6 above)
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: Resolved, for details, see [here](https://github.com/emdgroup/baybe/pull/86#discussion_r1464858527).

File: baybe/parameters/utils.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+            if param.name in specified_params:
+                raise ValueError(
+                    f"You provided several parameters with the name '{param.name}'."
+                )
+            specified_params[param.name] = param
+
+    # Try to find a parameter match for each dataframe column
+    parameters = []
+    for name, series in df.items():
+        assert isinstance(name, str)
```
Comment (AdrianSosic) [inline baybe/parameters/utils.py:57]: @Scienfitz: since you now also collected some experience already with mypy and in particular pandas, do you perhaps have a more clever way how to resolve the type ambiguity here? Had to insert the assert statements for type narrowing from `Hashable` (which is returned from `items()`) to `str`. Wouldn't now how else to make mypy aware that we expect the dataframe to hold string valued column names only, since pandas dataframes unfortunately cannot be typed. The alternative would be a `typing.cast`, which would just disable mypy's complains here, though the complains are valid!

Comment (Scienfitz) [reply baybe/parameters/utils.py:57]: uff hard to follow without seeing the original error and line, can you post?

in worst case a `cast` is fine

Comment (AdrianSosic) [reply baybe/parameters/utils.py:57]: Well, without the cast, two methods called in the body that expect a string raise a warning. For example, the call to `factory` gives:
image
So not sure what is the better approach. Cast has no runtime overhead, but assert actually checks that it's a string (which we have nowhere ensured, actually).

Comment (Scienfitz) [reply baybe/parameters/utils.py:57]: only way to avoid it then is to add dataframe valdiation, whcih seems overkill to me? assert with human-understandable error comment seems more reasonable to me then

Comment (AdrianSosic) [reply baybe/parameters/utils.py]: Ok, have done so. Will resolve. From my side, the PR is ready. Once you resolve the other thread, I'll proceed to merge 🚀

File: baybe/searchspace/core.py (hunk 8 above)
Comment (Scienfitz) [reply baybe/searchspace/core.py]: gucci
